src/analyses/apron/apronPriv.apron.ml

```diff
@@ -474,30 +474,183 @@ struct
+module Clustering12: ClusteringArg =
+struct
+  let generate gs =
+    List.cartesian_product gs gs
+    |> List.filter (fun (g1, g2) -> CilType.Varinfo.compare g1 g2 <= 0) (* filter flipped ordering, keep equals for next step *)
+    |> List.map (fun (g1, g2) -> [g1; g2]) (* if g1 = g2, then we get a singleton cluster *)
+end
+
+(** All clusters of size 2. *)
+module Clustering2: ClusteringArg =
+struct
+  let generate gs =
+    match gs with
+    | [_] -> [gs] (* use clusters of size 1 if only 1 variable *)
+    | _ ->
+      List.cartesian_product gs gs
+      |> List.filter (fun (g1, g2) -> CilType.Varinfo.compare g1 g2 < 0) (* filter flipped ordering, forbid equals for just clusters of size 2 *)
+      |> List.map (fun (g1, g2) -> [g1; g2])
+end
+
+(** All subset clusters. *)
+module ClusteringPower: ClusteringArg =
+struct
+  let generate gs =
+    gs
+    |> List.map (fun _ -> [true; false])
+    |> List.n_cartesian_product (* TODO: exponential! *)
+    |> List.map (fun bs ->
+      (* list globals where omega is true *)
+      List.fold_left2 (fun acc g b ->
+          if b then
+            g :: acc
+          else
+            acc
+        ) [] gs bs
+    )
+end
+
+(** One maximum cluster. *)
+module ClusteringMax: ClusteringArg =
+struct
+  let generate gs =
+    [gs]
+end
+
+
+module Cluster (ClusteringArg: ClusteringArg): ClusterArg =
+struct
+  open CommonPerMutex
+
+  module VS =
+  struct
+    include Printable.Std
+    include SetDomain.Make (CilType.Varinfo)
+  end
+  module LAD = MapDomain.MapBot (VS) (AD)
+
+  let filter_map' f m =
+    LAD.fold (fun k v acc ->
+        match f k v with
+        | Some (k', v') ->
+          LAD.add k' (AD.join (LAD.find k' acc) v') acc
+        | None ->
+          acc
+      ) m (LAD.empty ())
+
+  let keep_only_protected_globals ask m octs =
+    filter_map' (fun gs oct ->
+        (* must filter by protection to avoid later meeting with non-protecting *)
+        let gs' = VS.filter (is_protected_by ask m) gs in
+        if VS.is_empty gs' then
+          None
+        else
+          (* must restrict cluster down to protected (join) *)
+          Some (gs', keep_only_protected_globals ask m oct)
+      ) octs
+
+  let keep_global g octs =
+    filter_map' (fun gs oct ->
+        (* must filter by protection to avoid later meeting with non-protecting *)
+        if VS.mem g gs then (
+          let g_var = V.global g in
+          (* must restrict cluster down to m_g (join) *)
+          Some (VS.singleton g, AD.keep_vars oct [g_var])
+        )
+        else
+          None
+      ) octs
+
+  let lock local_m get_m =
+    let joined = LAD.join local_m get_m in
+    if M.tracing then M.traceli "apronpriv" "cluster12 lock:\n  local=%a\n  get=%a\n  joined=%a\n" LAD.pretty local_m LAD.pretty get_m LAD.pretty joined;
+    let r = LAD.fold (fun _ -> AD.meet) joined (AD.bot ()) in (* bot is top with empty env *)
+    if M.tracing then M.traceu "apronpriv" "-> %a\n" AD.pretty r;
+    r
```

Comment:
How does this take care of clusters that are too large initially, where the smaller subclusters are not tracked?
Consider e.g. at some mutex m, we initially considered it protecting for `{g,h,i}`, but later discovered that it only protects `{g,h}`:

In the case of Cluster2, we have clusters `{{g,h},{h,i}.{g,i}}`. Later, we will only publish to the cluster `{g,h}`. We will however still read things from the clusters `{h,i}` (restricted to `{h}`) and `{g,i}` (restricted to `{g}`). These clusters never get any new contributions after we have found out that `i` is not protected.

At the `lock(m)`, however, we join cluster-wise over all different threads (including those ones that are restricted to `{g}` and `{h}`), and then **meet** the result of the different clusters. Would this meet then not include the value for the clusters `{g}` and `{h}`, that does not soundly describe all values of `g` and `h`? Or am I missing something here?

Reply:
This probably still is why `cluster2` doesn't work completely. It's exactly the decreasing locksets problem that I've been having issues with. Not sure what the solution is here because at some point you do need to take the values from the larger lockset and have it propagated into the smaller one.

So far I tried to somehow do it at read time, like we've been doing for everything else, but maybe it's not fully possible here and there needs to be an explicit step for it when protecting locksets change.